### PR TITLE
Fix loading material diameter and GUID from XML file

### LINKED ISSUE #1886: 2.6.0beta Materials issues
Materials settings doesn't override some profile settings.
(works with inbuilt profiles, doesn't work with custom profiles)

I've created custom material and when selected -
- it doesn't override - hotted temp, filament diameter, retraction distance
- does override - bed temp, retraction speed, fan speed
- unable to test standby temp hence not configurable in print settings

Also regardless from the profile (custom or inbuilt) -
- in layer view mode with material colour selected it doesn't change preview to one in the Material config
- doesn't calculate cost of the print

OS - OSX 10.12.5 (clean install)

### PR DESCRIPTION
**Pull request by fieldOfView**
This PR fixes loading the material diameter from xml material files. Diameters were previously only shown, but not used as values.

Fixes #1886 

While I was at it, I also added loading the GUID into the material_guid setting. This setting would never be filled in.

### CODE REVIEW

**Comment by fieldOfView** [comment]
PR is against master, but this fix may have to be cherry-picked to 2.6 since it fixes a long-standing bug and 2.6 has a lot of material-related fixes in it so it fits with the "theme".

**Comment by Ghostkeeper** [comment]
Devs, see CURA-2822.

**Comment by Ghostkeeper** [comment]
CURA-2822 had a link to this Github issue as well: https://github.com/Ultimaker/Cura/issues/1308
I think we have some duplicates around.

File: plugins/XmlMaterialProfile/XmlMaterialProfile.py [hunk 1]
```diff
@@ -439,6 +439,8 @@ def deserialize(self, serialized):
         meta_data["base_file"] = self.id
         meta_data["status"] = "unknown"  # TODO: Add material verfication
 
+        global_setting_values = {}
```
Comment (BagelOrb) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Global?! You should be changing only the settings of the extruder in which this material is loaded!

This is either a bug or a rather unfortunate name for a variable.... Or is it?

Comment (LipuFei) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Unfortunately this was the name we used... See line 489.

Comment (fieldOfView) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Note that I moved this line of code from line 489, I did not introduce this variable name.

Comment (fieldOfView) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: "global" in this context is "global for all the containers that get created from a single xml file", not "global for all extrudertrains".

Comment (LipuFei) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: I think the "global" here means the settings in the "settings" section in the XML material file, which can applied to any machine. The settings in "settings/machine" sections are machine specific settings.

Comment (BagelOrb) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Hmmmm interesting..... ^^

Comment (nallath) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Global in this context should indeed be seen as "general"

**Comment by fieldOfView** [comment]
Also see https://github.com/Ultimaker/Cura/issues/1273

File: plugins/XmlMaterialProfile/XmlMaterialProfile.py (hunk 1 above)
Comment (LipuFei) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: The rest looks fine to me. @BagelOrb Do you still prefer to have this variable renamed?

Comment (BagelOrb) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Words like global, general, all, overall fail to convey enough meaning. They all convey that it is unbounded in the field of subject without saying what the field of subject is. So in contrast to printer specific settings these global settings are not bounded to any application of the material.

The problem is that variable names should be self explanatory, while 'global' only has meaning wrt 'printer specific'. 

I don't know. Maybe this is just a philosophical excerpt.

Comment (fieldOfView) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: I have changed the word "global" into "common". While this does not fix the entire philosophical issue of not specifying the context of the word global/general/common, it does take away the confusion over the term "global" which has such a strong connotation in Cura.

**Comment by BagelOrb** [comment]
Do I need to clean the cache or something?
I cannot get this to work...

**Comment by BagelOrb** [comment]
I'm testing like so:
- open material manager
- create new material
- change density and diameter
- close cura
- start cura
- check new material in the manager

File: plugins/XmlMaterialProfile/XmlMaterialProfile.py (hunk 1 above)
Comment (BagelOrb) [inline plugins/XmlMaterialProfile/XmlMaterialProfile.py]: Fair nuf

**Comment by BagelOrb** [comment]
With the above test case failing this PR doesn't really add much value... At least before this commit it's clear that it doesn't work. With this PR the user has no idea the value reset when he reopened Cura....

Please fix the above issue? Pretty Please? :pray:

**Comment by BagelOrb** [comment]
Also we have decided to merge this to 2.6, so either you or we will cherry-pick the commits into 2.6.

**Comment by fieldOfView** [comment]
> At least before this commit it's clear that it doesn't work.

Can't argue with that reasoning...  :weary:

It turns out setting metadata for xmlmaterialprofiles was broken, which meant that changes did not get saved. I have also fixed the issue that changing the diameter in the material view did not also change it in the settings.

This PR now fixes deserialising, changing and serialising xmlmaterialprofiles.

**Comment by BagelOrb** [comment]
I think I have disappearing custom materials with the new commits you made...

**Comment by fieldOfView** [comment]
Repro steps?

Note that if you change a 2.85 mm material to a 1.75 mm material, it it disappears because it is no longer suitable for the current machine.

**Comment by BagelOrb** [comment]
Could not deserialize any def.json file: 
`json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

**Comment by BagelOrb** [comment]
Never mind. I just ffed up :O

**Comment by BagelOrb** [comment]
Why is 1.75 mm material incompatible by default?

Does that hold for any machine?

**Comment by fieldOfView** [comment]
1.75mm was just an example

**Comment by BagelOrb** [comment]
I get that, but why is a material with a new diameter incompatible by default?

If I want to make a profile for the 1.2342535 mm filament I just bought I want to create a new profile and not have it disappear. Why would it disappear?

**Comment by fieldOfView** [comment]
I fixed the case where the user changes the diameter of the currently active material. This now properly refreshes the list of materials to a list of materials with the same approximate diameter. This is the logical cause of the material diameter filtering that was introduced before.

Material management was very broken. With this PR it is much less broken. Remaining usability problems should (IMHO) be logged as a new issue so they can be fixed later.

> Why would it disappear?

Because it is not compatible with your current printer. You bought the wrong material.

**Comment by BagelOrb** [comment]
Hah! Awesome!

**Comment by Ghostkeeper** [comment]
Still some weird update bugs going on. When creating a new material and setting the diameter to 1.75mm (while using UM3) it didn't disappear (because it was active). Activating a different material then causes the new custom material to disappear in the material manager but not in the drop-down selection box in the side panel.

However this does make it a lot more stable so I'm cherry-picking this into 2.6. Hope I'm still on time for the 2.6.0 release.

**Comment by Ghostkeeper** [comment]
Cherry-picked into 2.6.

**Comment by fieldOfView** [comment]
> When creating a new material and setting the diameter to 1.75mm (while using UM3) it didn't disappear (because it was active)

Going forward Cura should probably show an Ok/Cancel dialog when changing the diameter in such a way that the approximate_diameter changes. The use-case that it currently works well with is changing the diameter from 2.85 to eg 2.9.

**Comment by Ghostkeeper** [comment]
I'm working on a warning message that's being displayed when you change the diameter in such a way that it disappears. The warning message has an undo button. This should address that issue.

**Pull request closed without merging**